src/snapred/ui/workflow/ReductionWorkflow.py

```diff
@@ -1,40 +1,55 @@
 from snapred.backend.api.InterfaceController import InterfaceController
 from snapred.backend.dao import RunConfig, SNAPRequest
 from snapred.backend.log.logger import snapredLogger
+from snapred.meta.decorators.ExceptionToErrLog import ExceptionToErrLog
 from snapred.ui.view.ReductionView import ReductionView
 from snapred.ui.workflow.WorkflowBuilder import WorkflowBuilder
+from snapred.ui.workflow.WorkflowImplementer import WorkflowImplementer
 
 logger = snapredLogger.getLogger(__name__)
 
 
-class ReductionWorkflow:
+class ReductionWorkflow(WorkflowImplementer):
     def __init__(self, parent=None):
-        # create a tree of flows for the user to successfully execute diffraction calibration
-        # Calibrate     ->
-        # Assess        ->
-        # Save?         ->
         self.requests = []
         self.responses = []
         self.interfaceController = InterfaceController()
-        cancelLambda = None
-        if parent is not None and hasattr(parent, "close"):
-            cancelLambda = parent.close
 
         self._reductionView = ReductionView(parent=parent)
 
+        self._reductionView.runNumberField.editingFinished.connect(self._populatePixelMaskDropdown)
+
         self.workflow = (
-            WorkflowBuilder(cancelLambda=cancelLambda, parent=parent)
+            WorkflowBuilder(cancelLambda=self.resetWithPermission, parent=parent)
             .addNode(
                 self._triggerReduction,
                 self._reductionView,
                 "Reduction",
             )
             .build()
         )
+        self.workflow.presenter.setResetLambda(self.reset)
+
+    @ExceptionToErrLog
+    def _populatePixelMaskDropdown(self):
+        runNumbers = self._reductionView.runNumberField.text().split(",")
+        useLiteMode = self._reductionView.litemodeToggle.field.getState()  # noqa: F841
+
+        self._reductionView.litemodeToggle.setEnabled(False)
+        self._reductionView.pixelMaskDropdown.setEnabled(False)
+
+        for runNumber in runNumbers:
+            try:
+                self.request(path="reduction/hasState", payload=runNumber).data
+            except Exception as e:  # noqa: BLE001
+                print(e)
```

Comment:
Why is this being handled inside this function for pixel mask drop down, and not another function?

Reply:
I was just following the template laid out within Diffcal workflow. Figured it would be good for consistency.